Add server support for updating properties in catalog and schema

tlm365 opened the pull request:
**PR Checklist**

- [x] A description of the changes is added to the description of this PR.
- [x] If there is a related issue, make sure it is linked to this PR.
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added or modified a feature, documentation in `docs` is updated

**Description of changes**
Closes #175.
<!-- Please state what you've changed and how it might affect the users. -->

tlm365 commented:
> [error] Test io.unitycatalog.cli.catalog.CliCatalogCRUDTest.testCatalogCRUD[0] failed: org.opentest4j.AssertionFailedError:

The issue will likely be handled by another PR, related to #175.

@vikrantpuppala Could you please take a look?

vikrantpuppala commented on server/src/main/java/io/unitycatalog/server/persist/dao/CatalogInfoDAO.java:
```diff
@@ -28,6 +30,10 @@ public class CatalogInfoDAO {
 
+    @Column(name = "properties")
+    @JdbcTypeCode(SqlTypes.JSON)
+    private Map<String, String> properties;
```
The way properties work is that they are stored in a different table constructed by PropertyInfoDAO, we should use the same structure instead of adding a new column. These are references of how it is done [created](https://github.com/unitycatalog/unitycatalog/blob/main/server/src/main/java/io/unitycatalog/server/persist/CatalogRepository.java#L53) and [deleted](https://github.com/unitycatalog/unitycatalog/blob/main/server/src/main/java/io/unitycatalog/server/persist/CatalogRepository.java#L194).

vikrantpuppala commented on server/src/test/java/io/unitycatalog/server/base/catalog/BaseCatalogCRUDTest.java:
```diff
@@ -1,37 +1,39 @@
+    protected void assertCatalog(CatalogInfo catalogInfo, String name, String comment, Map<String, String> properties) {
+        assertThat(catalogInfo.getName()).isEqualTo(name);
+        assertThat(catalogInfo.getComment()).isEqualTo(comment);
+        assertThat(catalogInfo.getProperties()).isEqualTo(properties);
```
Let us not add this assert right now so that the tests do not fail, we will fix it with the TODO on line 25

tlm365 replied:
@vikrantpuppala thanks so much for reviewing, but it seems that pending the properties assertion right now can't prevent cli-related errors

vikrantpuppala commented on server/src/test/java/io/unitycatalog/server/base/catalog/BaseCatalogCRUDTest.java:
```diff
@@ -1,35 +1,36 @@
-    Assert.assertEquals(name, catalogInfo.getName());
-    Assert.assertEquals(comment, catalogInfo.getComment());
-    Assert.assertNotNull(catalogInfo.getCreatedAt());
+    assertThat(catalogInfo.getName()).isEqualTo(name);
```
The [tests seem to be failing](https://github.com/unitycatalog/unitycatalog/actions/runs/9889636546/job/27316141620?pr=190#step:4:2335) at this assertion? why is that?

tlm365 replied:
@vikrantpuppala hmm, that's werd to me as well, initially I thought it depends on the cli implementation because the error log is related to the cli:
>[error] Test io.unitycatalog.**cli.catalog.CliCatalogCRUDTest**.testCatalogCRUD[0] failed: java.lang.NullPointerException: null.

But when I tried to find the root cause, I found 2 strange things:
(1) It's fine if we run testOnly: `build/sbt 'testOnly io.unitycatalog.cli.catalog.CliCatalogOperations'` but error if run with: `build/sbt test`.
(2) `NullPointerException` (same as error above) also threw at `create catalog`'s test-case when comment // [this line](https://github.com/unitycatalog/unitycatalog/blob/ed2ad4af6fbd663e45a0271c2a750d9e2acab9eb/examples/cli/src/test/java/io/unitycatalog/cli/TestUtils.java#L23) (just testing, I don't quite understand the intention of this error handling). But it doesn't work with the new added `update catalog`'s test-case.

I will dig deeper into this issue when I have more time this weekend.

VillePuuska replied:
This seems to be because of updating properties not implemented for the CLI combined with how the CLI validates input; specifically how no-ops are handled. The CLI sees [here](https://github.com/unitycatalog/unitycatalog/blob/bb8dd4fddee8390043e4133c767165d0baa1050b/examples/cli/src/main/java/io/unitycatalog/cli/CatalogCli.java#L77-L85) that there are no supported optional parameters so it does not return the info about the existing catalog and instead prints a help message. If you comment that part out the CLI passes the command to the server which returns the catalog info and the test then passes.

You can work around this by getting the catalog info after the update with `getCatalog` separately like is done for the empty update test [here](https://github.com/unitycatalog/unitycatalog/blob/bb8dd4fddee8390043e4133c767165d0baa1050b/server/src/test/java/io/unitycatalog/server/base/catalog/BaseCatalogCRUDTest.java#L60-L61). So the test for updating properties could be e.g.:

    // Update catalog properties without updating name and comment
    System.out.println("Testing update catalog: changing properties..");
    UpdateCatalog updateCatalog3 = new UpdateCatalog().properties(NEW_PROPERTIES);
    CatalogInfo updatedCatalogInfo3 =
        catalogOperations.updateCatalog(CATALOG_NEW_NAME, updateCatalog3);
    CatalogInfo catalogInfo4 = catalogOperations.getCatalog(CATALOG_NEW_NAME);
    assertCatalog(catalogInfo4, CATALOG_NEW_NAME, CATALOG_NEW_COMMENT);

vikrantpuppala reviewed (commented):
thanks for addressing comments, just need to get the unit tests sorted.

vikrantpuppala reviewed (approved):
@haogang to merge

vikrantpuppala commented:
@haogang could you check this as well?

rameshchandra commented:
@tlm365 can you rebase this PR on the latest changes from main?

vikrantpuppala commented on server/src/main/java/io/unitycatalog/server/persist/CatalogRepository.java:
```diff
@@ -159,6 +161,13 @@ && getCatalogDAO(session, updateCatalog.getNewName()) != null) {
+        }
         catalogInfoDAO.setUpdatedAt(new Date());
         session.merge(catalogInfoDAO);
         tx.commit();
```
The CatalogInfo returned below would not have the properties, can we use the `PropertyRepository` to add that?

vikrantpuppala commented on server/src/main/java/io/unitycatalog/server/persist/SchemaRepository.java:
```diff
@@ -208,6 +210,13 @@ public SchemaInfo updateSchema(String fullName, UpdateSchema updateSchema) {
+        }
         schemaInfo.setUpdatedAt(new Date());
         session.merge(schemaInfo);
         tx.commit();
```
We should attach properties inside `convertFromDAO` below

vikrantpuppala commented on server/src/main/java/io/unitycatalog/server/persist/CatalogRepository.java:
```diff
@@ -55,6 +55,8 @@ public CatalogInfo addCatalog(CreateCatalog createCatalog) {
         PropertyDAO.from(catalogInfo.getProperties(), catalogInfoDAO.getId(), Constants.CATALOG)
             .forEach(session::persist);
+        RepositoryUtils.attachProperties(
+            catalogInfo, catalogInfo.getId(), Constants.CATALOG, session);
```
These would already be present from [here](https://github.com/unitycatalog/unitycatalog/pull/190/files#diff-7c78786fc147d2887740af532f7538cd50e04960aa35c7a77e1c90cbb210c096R46), we do not need this change

vikrantpuppala commented on server/src/main/java/io/unitycatalog/server/persist/CatalogRepository.java:
```diff
@@ -159,6 +163,17 @@ && getCatalogDAO(session, updateCatalog.getNewName()) != null) {
+
+          CatalogInfo catalogInfo = catalogInfoDAO.toCatalogInfo();
+          RepositoryUtils.attachProperties(
+              catalogInfo, catalogInfo.getId(), Constants.CATALOG, session);
```
At line 180, we are returning catalogInfoDAO.toCatalogInfo(), so this change would have no effect, we should remove this and just do `RepositoryUtils.attachProperties(
              catalogInfoDAO.toCatalogInfo(), catalogInfo.getId(), Constants.CATALOG, session);` at line 180

vikrantpuppala replied:
Since we want the properties irrespective of `if (!updateCatalog.getProperties().isEmpty())`

vikrantpuppala commented on server/src/main/java/io/unitycatalog/server/persist/SchemaRepository.java:
```diff
@@ -208,10 +210,19 @@ public SchemaInfo updateSchema(String fullName, UpdateSchema updateSchema) {
         tx.commit();
-        return convertFromDAO(schemaInfo, fullName);
+        SchemaInfo schema = convertFromDAO(schemaInfo, fullName);
+        return RepositoryUtils.attachProperties(
```
can we move this into the convertFromDAO function instead?

vikrantpuppala commented on server/src/test/java/io/unitycatalog/server/base/catalog/BaseCatalogCRUDTest.java:
```diff
@@ -55,24 +55,30 @@ public void testCatalogCRUD() throws ApiException {
+    UpdateCatalog updateCatalog3 = new UpdateCatalog().properties(NEW_PROPERTIES);
+    catalogOperations.updateCatalog(CATALOG_NEW_NAME, updateCatalog3);
+    CatalogInfo updatedCatalogInfo3 = catalogOperations.getCatalog(CATALOG_NEW_NAME);
+    assertCatalog(updatedCatalogInfo3, CATALOG_NEW_NAME, CATALOG_NEW_COMMENT);
```
But assertCatalog do not assert properties so we aren't event checking the property update? We also would need to address the TODO on line 22

vikrantpuppala commented on server/src/test/java/io/unitycatalog/server/base/schema/BaseSchemaCRUDTest.java:
```diff
@@ -84,6 +83,16 @@ public void testSchemaCRUDL() throws ApiException {
+    assertThat(retrievedSchemaInfo3.getName()).isEqualTo(TestUtils.SCHEMA_NEW_NAME);
+    assertThat(retrievedSchemaInfo3.getComment()).isEqualTo(TestUtils.SCHEMA_COMMENT);
+    assertThat(retrievedSchemaInfo3.getFullName()).isEqualTo(TestUtils.SCHEMA_NEW_FULL_NAME);
+    assertThat(retrievedSchemaInfo3.getUpdatedAt()).isNotNull();
```
We also need to assert properties?

vikrantpuppala reviewed (changes requested):
thank you for updating, needs some fixes

vksx commented on server/src/main/java/io/unitycatalog/server/persist/CatalogRepository.java:
```diff
@@ -159,6 +163,17 @@ && getCatalogDAO(session, updateCatalog.getNewName()) != null) {
         if (updateCatalog.getComment() != null) {
           catalogInfoDAO.setComment(updateCatalog.getComment());
         }
+        if (!updateCatalog.getProperties().isEmpty()) {
```
Please use both `null` and `isEmpty()` checks for properties' absence for consistency.

vksx commented on server/src/main/java/io/unitycatalog/server/persist/SchemaRepository.java:
```diff
@@ -208,10 +210,19 @@ public SchemaInfo updateSchema(String fullName, UpdateSchema updateSchema) {
         if (updateSchema.getNewName() != null) {
           schemaInfo.setName(updateSchema.getNewName());
         }
+        if (!updateSchema.getProperties().isEmpty()) {
```
Please use both `null` and `isEmpty()` checks for properties' absence for consistency.

vksx commented on server/src/main/java/io/unitycatalog/server/persist/SchemaRepository.java:
```diff
@@ -197,7 +197,9 @@ public SchemaInfo updateSchema(String fullName, UpdateSchema updateSchema) {
+            && updateSchema.getNewName() == null
+            && (updateSchema.getProperties() == null || updateSchema.getProperties().isEmpty())) {
           tx.rollback();
           return convertFromDAO(schemaInfo, fullName);
```
We should rename `schemaInfo` to `schemaInfoDAO` at line [190](https://github.com/unitycatalog/unitycatalog/pull/190/files#diff-3220d6f332348a34e53ba25cec22a52726afa71cf8f7d4fc31c55b1aee34477bR190) to avoid any confusion and correspondingly update all it's occurrences as follows:
```suggestion
          return convertFromDAO(schemaInfoDAO, fullName);
```

vksx commented on server/src/test/java/io/unitycatalog/server/base/catalog/BaseCatalogCRUDTest.java:
```diff
@@ -55,24 +55,30 @@ public void testCatalogCRUD() throws ApiException {
+    // Update catalog properties without updating name and comment
+    System.out.println("Testing update catalog: changing properties..");
+    UpdateCatalog updateCatalog3 = new UpdateCatalog().properties(NEW_PROPERTIES);
+    catalogOperations.updateCatalog(CATALOG_NEW_NAME, updateCatalog3);
```
Lines 78 and 79 can be turned into a single line for more clean code as follows:
```suggestion
CatalogInfo updatedCatalogInfo3 =
        catalogOperations.updateCatalog(CATALOG_NEW_NAME, updateCatalog3);
```
You may refer [this](https://github.com/unitycatalog/unitycatalog/pull/338/files#diff-19017c54d61c82634f9611c323bf5bcf155d2e400a0aa21699b7046580651debR85) as well for other changes of this file - including the TODO on line 22.

vksx commented on server/src/test/java/io/unitycatalog/server/base/schema/BaseSchemaCRUDTest.java:
```diff
@@ -84,6 +83,16 @@ public void testSchemaCRUDL() throws ApiException {
+    System.out.println("Testing update schema: changing properties..");
+    UpdateSchema updateSchema3 = new UpdateSchema().properties(TestUtils.NEW_PROPERTIES);
+    schemaOperations.updateSchema(TestUtils.SCHEMA_NEW_FULL_NAME, updateSchema3);
+    SchemaInfo retrievedSchemaInfo3 = schemaOperations.getSchema(TestUtils.SCHEMA_NEW_FULL_NAME);
```
Rather than changing the naming convention - please rename `retrievedSchemaInfo` to `updatedSchemaInfo3` and change the name of the occurrence at line 99 to `updatedSchemaInfo4` for consistency. 

Also, for other changes related to this file including the TODO - you may refer [this](https://github.com/unitycatalog/unitycatalog/pull/338/files#diff-aaa8d031b8b88f3e944c9409431246ca3b2ddc9a9b438415b067faa649ed3505).

vksx commented:
@tlm365 Thanks for this PR! I was [unaware](https://github.com/unitycatalog/unitycatalog/pull/338#issuecomment-2289752638) that you were already working on this so I also made the server changes. You may refer to #338 for more details, if required, to resolve the review comments.

tlm365 commented:
@vksx Thanks for reviewing, noted. Looks like this PR will have some conflicts with #338 (quick look at TestUtils.NEW_PROPERTIES), I will finish this PR after #338 is merged.

vksx commented:
@tlm365 I have [reverted](https://github.com/unitycatalog/unitycatalog/pull/338/commits/c568eb528d92a0c045cc4a5fce9a4bf190d3a81f) the server file changes to keep my changes in the scope of #176. Please go ahead with your changes - there will not be any conflicts now.

vikrantpuppala commented:
@tlm365 a gentle ping on this :) would like to close this as soon as possible, thanks in advance!

tlm365 commented:
> @tlm365 a gentle ping on this :) would like to close this as soon as possible, thanks in advance!

@vikrantpuppala @vksx Thanks for your review and patience, I updated some minor fixes and added properties assertion (1) (category & schema) to the unit test. But as I mentioned before, I really think CLI PR and Server PR need proper interaction for (1) to solve the problem.

vikrantpuppala commented:
thanks @tlm365, that makes sense. @vksx what's your opinion on this one? I think there's a bit of confusion in trying to do these separately, should we do all of it in one PR? maybe we can take Tai's commits from this PR into your PR?

vksx commented:
@vikrantpuppala Sure, we can do that.
